registry/kubernetes/registry.go

```diff
@@ -0,0 +1,239 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kubernetes
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"strings"
+	"sync"
+	"time"
+)
+
+import (
+	"github.com/dubbogo/getty"
+	"github.com/dubbogo/gost/net"
+	perrors "github.com/pkg/errors"
+	k8s "k8s.io/client-go/kubernetes"
+)
+
+import (
+	"github.com/apache/dubbo-go/common"
+	"github.com/apache/dubbo-go/common/constant"
+	"github.com/apache/dubbo-go/common/extension"
+	"github.com/apache/dubbo-go/common/logger"
+	"github.com/apache/dubbo-go/registry"
+	"github.com/apache/dubbo-go/remoting/kubernetes"
+)
+
+var (
+	processID = ""
+	localIP   = ""
+)
+
+const (
+	Name         = "kubernetes"
+	ConnDelay    = 3
+	MaxFailTimes = 15
+)
+
+func init() {
+	processID = fmt.Sprintf("%d", os.Getpid())
+	localIP, _ = gxnet.GetLocalIP()
+	extension.SetRegistry(Name, newKubernetesRegistry)
+}
+
+type kubernetesRegistry struct {
+	registry.BaseRegistry
+	cltLock        sync.RWMutex
+	client         *kubernetes.Client
+	listenerLock   sync.Mutex
+	listener       *kubernetes.EventListener
+	dataListener   *dataListener
+	configListener *configurationListener
+}
+
+func (r *kubernetesRegistry) Client() *kubernetes.Client {
+	r.cltLock.RLock()
+	client := r.client
+	r.cltLock.RUnlock()
+	return client
+}
+func (r *kubernetesRegistry) SetClient(client *kubernetes.Client) {
+	r.cltLock.Lock()
+	r.client = client
+	r.cltLock.Unlock()
+}
+
+func (r *kubernetesRegistry) CloseAndNilClient() {
+	r.client.Close()
+	r.client = nil
```

Comment:
```suggestion
	r.cltLock.Lock()
	client:=r.client
	r.client=nil
	r.cltLock.Unlock()
	client.Close()
```

Reply:
The invoker has locked. So it's not neccessary add lock here.

Reply:
i argee.